### DISCUSSION TITLE
PyTorch YoloV5 but with different results (Under the same environement).

### DISCUSSION THREAD

**Discussion by lujingyang1029** [ultralytics/yolov5#7627] [Q&A]
1) first is from loading the model from torch.hub.load()

2) Second is from python detect.py

As instructed from the yolov5 git, the two ways should output the identity results. It turns out not. Please help fix. What lead to this problem. For some certain object, they can both detect it but with different confidence level (0.4 vs 0.7). 

Thanks.

**Comment by glenn-jocher** [comment]
@lujingyang1029 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃

**Reply by lujingyang1029** [reply]
1. Using  torch.hub.load(), Run the code as follows,

import cv2
import torch
from PIL import Image
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')
im1 = cv2.imread('Path/To/bus.jpg')
im2 = Image.open('Path/To/zidane.jpg')  # PIL image
imgs = [im1, im2]
result = model(imgs, size = 640)
[result.show()](url)
print(result.pandas().xyxy)

2. Using python detect.py
python detect.py --source path/to/yolov5/data/images
![tmplyqf16p5](https://user-images.githubusercontent.com/16713826/165859209-65c50ce3-e1e5-4e30-8cdc-3bc224f13529.PNG)
![bus](https://user-images.githubusercontent.com/16713826/165859222-0d070a99-b301-4512-bd85-cb1576e77e7e.jpg)

**Reply by lujingyang1029** [reply]
The bus and the person at the right corner, the confidence level is different. It is even larger for other custom dataset, which i think is not considered here.

**Reply by glenn-jocher** [reply]
@lujingyang1029 your cv2 usage is incorrect, follow **PyTorch Hub** tutorial:

## YOLOv5 Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://github.com/ultralytics/yolov5/wiki/Tips-for-Best-Training-Results)&nbsp; ☘️  RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Roboflow for Datasets, Labeling, and Active Learning](https://github.com/ultralytics/yolov5/issues/4975)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [TFLite, ONNX, CoreML, TensorRT Export](https://github.com/ultralytics/yolov5/issues/251) 🚀
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [Architecture Summary](https://github.com/ultralytics/yolov5/issues/6998)&nbsp; ⭐ NEW

Good luck 🍀 and let us know if you have any other questions!

**Comment by solisoares** [comment]
How things go unnoticed. 

@lujingyang1029 ,I know it's been a while but I'd like to add that, even though @glenn-jocher did not point exactly what was wrong with your cv2 usage, it is lacking an additional "BGR to RGB" conversion as mentioned in this [line](https://github.com/ultralytics/yolov5/issues/36#:~:text=im2%20%3D%20cv2.imread(%27bus.jpg%27)%5B...%2C%20%3A%3A%2D1%5D%20%20%23%20OpenCV%20image%20(BGR%20to%20RGB)) in the PyTorch Hub Tutorial. 

It was not the first time I visited this tutorial, but sometimes we just don't see what is in front of our eyes. I was experiencing the same problem as you. To be precise, I was detecting frames read from a opencv capture, and all that was missing was this conversion to RGB I mentioned.

As stated in the source code snippet below, when loading an image with opencv we need to do this conversion.
```python
'''File: https://github.com/ultralytics/yolov5/blob/master/models/common.py'''

@smart_inference_mode()
def forward(self, ims, size=640, augment=False, profile=False):
    # Inference from various sources. For size(height=640, width=1280), RGB images example inputs are:
    #   file:        ims = 'data/images/zidane.jpg'  # str or PosixPath
    #   URI:             = 'https://ultralytics.com/images/zidane.jpg'
--> #   OpenCV:          = cv2.imread('image.jpg')[:,:,::-1]  # HWC BGR to RGB x(640,1280,3)
    #   PIL:             = Image.open('image.jpg') or ImageGrab.grab()  # HWC x(640,1280,3)
    #   numpy:           = np.zeros((640,1280,3))  # HWC
    #   torch:           = torch.zeros(16,3,320,640)  # BCHW (scaled to size=640, 0-1 values)
    #   multiple:        = [Image.open('image1.jpg'), Image.open('image2.jpg'), ...]  # list of images
```

I really tried not to mess up with the source code for my needs, but after this, I compared same frames against the script and torch hub detections and concluded they were equal, fixing my problem.

**Reply by glenn-jocher** [reply]
Yes, cv2 loads in BGR, YOLOv5 expect RGB inputs.
